test/e2e/reconciler/apiexport/apiexport_controller_test.go

```diff
@@ -119,3 +123,70 @@ func TestRequeueWhenIdentitySecretAdded(t *testing.T) {
 		return gotHash == "e9cee71ab932fde863338d08be4de9dfe39ea049bdafb342ce659ec5450b69ae"
 	}, wait.ForeverTestTimeout, 100*time.Millisecond, "expected identity valid condition and matching hash, got hash %s", gotHash)
 }
+
+func TestVirtualWorkspacesAssigned(t *testing.T) {
+	t.Parallel()
+
+	server := framework.SharedKcpServer(t)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	orgClusterName := framework.NewOrganizationFixture(t, server)
+	workspaceClusterName := framework.NewWorkspaceFixture(t, server, orgClusterName)
+	t.Logf("Running test in cluster %s", workspaceClusterName)
+
+	cfg := server.BaseConfig(t)
+
+	kcpClusterClient, err := clientset.NewForConfig(cfg)
+	require.NoError(t, err, "error creating kcp cluster client")
+
+	apiExport := &apisv1alpha1.APIExport{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "my-export",
+		},
+		Spec: apisv1alpha1.APIExportSpec{},
+	}
+
+	t.Logf("Creating APIExport")
+	apiExportClient := kcpClusterClient.ApisV1alpha1().APIExports()
+
+	_, err = apiExportClient.Create(logicalcluster.WithCluster(ctx, workspaceClusterName), apiExport, metav1.CreateOptions{})
+	require.NoError(t, err, "error creating APIExport")
+
+	t.Logf("Verifying if the APIExport gets VirtualWorkspaces assigned")
+	var details string
+	require.Eventually(t, func() bool {
+		export, err := apiExportClient.Get(logicalcluster.WithCluster(ctx, workspaceClusterName), apiExport.Name, metav1.GetOptions{})
+		if err != nil {
+			t.Logf("failed getting the ApiExport %q, err %v", apiExport.Name, err)
+			return false
+		}
+		clusterWorkspaceShards, err := kcpClusterClient.TenancyV1alpha1().ClusterWorkspaceShards().List(logicalcluster.WithCluster(ctx, tenancyv1alpha1.RootCluster), metav1.ListOptions{})
+		if err != nil {
+			t.Logf("failed getting ClusterWorkspaceShards, err %v", err)
+			return false
+		}
+
+		clusterWorkspaceShardVirtualWorkspaceURLs := sets.NewString()
+		for _, s := range clusterWorkspaceShards.Items {
+			if len(s.Spec.VirtualWorkspaceURL) == 0 {
+				t.Logf("%q shard hasn't had assigned a virtual workspace URL", s.Name)
+				return false
+			}
+			clusterWorkspaceShardVirtualWorkspaceURLs.Insert(s.Spec.VirtualWorkspaceURL)
+		}
+		exportVirtualWorkspaceURLS := sets.NewString()
+		for _, s := range export.Status.VirtualWorkspaces {
+			actualURL, err := url.Parse(s.URL)
+			if err != nil {
+				t.Log(err)
+				return false
+			}
+			exportVirtualWorkspaceURLS.Insert(fmt.Sprintf("%s://%s", actualURL.Scheme, actualURL.Host))
+		}
+		details = fmt.Sprintf("clusterWorkspaceShardVirtualWorkspaceURLs %v, exportVirtualWorkspaceURLS %v", clusterWorkspaceShardVirtualWorkspaceURLs, exportVirtualWorkspaceURLS)
```

Comment:
Can we use `cmp.Diff()` or `.Difference()` on these sets? I imagine if this fails the first thing someone will be doing is checking to see what the difference is :)